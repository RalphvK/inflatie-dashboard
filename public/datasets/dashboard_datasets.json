{
  "cpi": [{"Bestedingscategorieen_code":"T001112","Bestedingscategorieen_title":"000000 Alle bestedingen","Perioden_code":"2021MM07","Perioden_title":"2021 juli","Measure":"Jaarmutatie CPI","Unit":"%","Value":"1.4"},{"Bestedingscategorieen_code":"T001112","Bestedingscategorieen_title":"000000 Alle bestedingen","Perioden_code":"2021MM08","Perioden_title":"2021 augustus","Measure":"Jaarmutatie CPI","Unit":"%","Value":"2.4"},{"Bestedingscategorieen_code":"T001112","Bestedingscategorieen_title":"000000 Alle bestedingen","Perioden_code":"2021MM09","Perioden_title":"2021 september","Measure":"Jaarmutatie CPI","Unit":"%","Value":"2.7"},{"Bestedingscategorieen_code":"T001112","Bestedingscategorieen_title":"000000 Alle bestedingen","Perioden_code":"2021MM10","Perioden_title":"2021 oktober","Measure":"Jaarmutatie CPI","Unit":"%","Value":"3.4"},{"Bestedingscategorieen_code":"T001112","Bestedingscategorieen_title":"000000 Alle bestedingen","Perioden_code":"2021MM11","Perioden_title":"2021 november","Measure":"Jaarmutatie CPI","Unit":"%","Value":"5.2"},{"Bestedingscategorieen_code":"T001112","Bestedingscategorieen_title":"000000 Alle bestedingen","Perioden_code":"2021MM12","Perioden_title":"2021 december","Measure":"Jaarmutatie CPI","Unit":"%","Value":"5.7"},{"Bestedingscategorieen_code":"T001112","Bestedingscategorieen_title":"000000 Alle bestedingen","Perioden_code":"2022MM01","Perioden_title":"2022 januari","Measure":"Jaarmutatie CPI","Unit":"%","Value":"6.4"},{"Bestedingscategorieen_code":"T001112","Bestedingscategorieen_title":"000000 Alle bestedingen","Perioden_code":"2022MM02","Perioden_title":"2022 februari","Measure":"Jaarmutatie CPI","Unit":"%","Value":"6.2"},{"Bestedingscategorieen_code":"T001112","Bestedingscategorieen_title":"000000 Alle bestedingen","Perioden_code":"2022MM03","Perioden_title":"2022 maart","Measure":"Jaarmutatie CPI","Unit":"%","Value":"9.7"},{"Bestedingscategorieen_code":"T001112","Bestedingscategorieen_title":"000000 Alle bestedingen","Perioden_code":"2022MM04","Perioden_title":"2022 april","Measure":"Jaarmutatie CPI","Unit":"%","Value":"9.6"},{"Bestedingscategorieen_code":"T001112","Bestedingscategorieen_title":"000000 Alle bestedingen","Perioden_code":"2022MM05","Perioden_title":"2022 mei","Measure":"Jaarmutatie CPI","Unit":"%","Value":"8.8"},{"Bestedingscategorieen_code":"T001112","Bestedingscategorieen_title":"000000 Alle bestedingen","Perioden_code":"2022MM06","Perioden_title":"2022 juni","Measure":"Jaarmutatie CPI","Unit":"%","Value":"8.6"},{"Bestedingscategorieen_code":"T001112","Bestedingscategorieen_title":"000000 Alle bestedingen","Perioden_code":"2022MM07","Perioden_title":"2022 juli*","Measure":"Jaarmutatie CPI","Unit":"%","Value":"10.3"}]
}