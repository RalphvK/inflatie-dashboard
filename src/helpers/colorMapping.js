import { THEME } from '@/config/constants.js';

/**
 * Function that iterates through array and returns first color where value is large than min.
 * Mapping is provided as array, with minimum values from high to low. Example:
 * [
 *   { min: 4, color: 'red' },
 *   { min: 0, color: 'green' }
 * ]
 * The example above will return 'red' if the value is >= 4, and 'green' if >= 0.
 * If the value is lower than 0, the fallback (param 3) would be returned.
 * 
 * @param {Number} measureValue value to color for
 * @param {Array<Object>} colorArray array of objects like: { min: 4, color: 'red' }
 * @param {String} belowLast color below colorArray's last element's min value
 * @returns {String}
 */
export function colorMappingDesc(measureValue, colorArray = null, fallback = THEME.neutral) {
  // default color mapping for CPI
  if (!Array.isArray(colorArray)) {
    colorArray = mapping_YearOnYear;
  }
  // iterate through mapping and return first that is lower than value
  for (let i = 0; i < colorArray.length; i++) {
    if (measureValue >= colorArray[i].min) {
      return colorArray[i].color;
    }
  }
  return fallback;
}

export const mapping_neutral = [
  {min: 6, color: THEME.neutral},
  {min: 3, color: THEME.neutral},
  {min: 1, color: THEME.neutral},
  {min: 0, color: THEME.neutral},
  {min: -Infinity, color: THEME.neutral}
];

// inflation year-on-year color scale (considers deflation a bad thing)
export const mapping_YearOnYear_Inflation = [
  {min: 6, color: THEME.danger},
  {min: 3, color: THEME.warning},
  {min: 1, color: THEME.success},
  {min: 0, color: THEME.warning},
  {min: -Infinity, color: THEME.danger}
];

// default year-on-year color scale (considers deflation a warning)
export const mapping_YearOnYear = [
  {min: 6, color: THEME.danger},
  {min: 3, color: THEME.warning},
  {min: 1, color: THEME.success},
  {min: 0, color: THEME.success},
  {min: -Infinity, color: THEME.warning}
];

// default year-on-year color scale (considers deflation a good thing)
export const mapping_YearOnYear_Energy = [
  {min: 6, color: THEME.danger},
  {min: 3, color: THEME.warning},
  {min: 1, color: THEME.success},
  {min: 0, color: THEME.success},
  {min: -Infinity, color: THEME.success}
];

// color scale for year-on-year income change (eg. wages)
export const mapping_YearOnYear_Income = [
  {min: 6, color: THEME.success},
  {min: 3, color: THEME.hopeful},
  {min: 2, color: THEME.neutral},
  {min: 1, color: THEME.warning},
  {min: -Infinity, color: THEME.danger}
];

// PPT color mapping for when YearOnYear figure is positive
// (in this scenario PPT descrease means growth is slowing down)
export const mapping_PptMonth_PositiveYoY = [
  {min: 2, color: THEME.danger},
  {min: 0, color: THEME.danger},
  {min: -3, color: THEME.hopeful},
  {min: -20, color: THEME.green},
  {min: -Infinity, color: THEME.warning}
];

// PPT color mapping for income change (eg. wages)
export const mapping_PptMonth_Income = [
  {min: 2, color: THEME.success},
  {min: 1, color: THEME.hopeful},
  {min: 0.2, color: THEME.hopeful},
  {min: 0, color: THEME.warning},
  {min: -2, color: THEME.warning},
  {min: -Infinity, color: THEME.danger}
];