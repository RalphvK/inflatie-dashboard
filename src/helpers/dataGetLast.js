import { last } from 'lodash';
import { SentientPercentage } from '@/helpers/SentientNumber.js';

export function dataGetLast(dataset, valueColumn = 'Value')
{
  let lastRow = last(dataset) ?? null;
  return lastRow
    ? new SentientPercentage(lastRow[valueColumn])
    : null
}