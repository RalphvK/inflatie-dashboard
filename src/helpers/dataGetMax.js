import { isUndefined, max as _max, toNumber } from "lodash";

export function dataGetMax(valueSeries = []) {
  let numberSeries = [];
  // convert each element to number
  for (let i = 0; i < valueSeries.length; i++) {
    numberSeries[i] = toNumber(valueSeries[i]);
  }
  // get max
  let max = _max(numberSeries);
  return isUndefined(max) ? 0 : Math.ceil(max);
}