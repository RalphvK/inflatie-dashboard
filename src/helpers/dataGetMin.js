import { isUndefined, min as _min, toNumber } from "lodash";

export function dataGetMin(valueSeries = []) {
  let numberSeries = [];
  // convert each element to number
  for (let i = 0; i < valueSeries.length; i++) {
    numberSeries[i] = toNumber(valueSeries[i]);
  }
  // get min
  let min = _min(numberSeries);
  return isUndefined(min) ? 0 : Math.ceil(min);
}