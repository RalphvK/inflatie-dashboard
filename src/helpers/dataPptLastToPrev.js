import { round } from 'lodash';
import { SentientPercentagePoint } from '@/helpers/SentientNumbers.js';

/**
 * Function to calculate percentage point change of last row to previous row.
 * Result is returned as either a SentienPercentagePoint instance, or string '-' if empty.
 * @param {Array<Object>} dataset Array of records in dataset, as stored in the .json file
 * @param {String} valueColumn name of the column to use for calculating PPT to last month
 * @returns {Object|String}
 */
export function dataPptLastToPrev(dataset = null, valueColumn = 'Value') {
  // check if data is present and more than 1 element
  if (Array.isArray(dataset) && dataset.length > 0) {
    let lastValue = dataset[dataset.length - 1][valueColumn];
    let prevValue = dataset[dataset.length - 2][valueColumn];
    return new SentientPercentagePoint(round(lastValue - prevValue, 1));
  } else {
    return '-';
  }
}