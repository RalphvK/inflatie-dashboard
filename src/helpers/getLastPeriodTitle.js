import { extractLongMonth, extractYear } from "@/helpers/extractMonthYear.js";

export function getLastPeriodTitle_MY(rows, columnName = 'Perioden_title')
{
  if (typeof rows === 'undefined' || !Array.isArray(rows)) {
    return null;
  }
  const last = rows[rows.length - 1][columnName];
  return [extractLongMonth(last), extractYear(last)].join(' ');
}