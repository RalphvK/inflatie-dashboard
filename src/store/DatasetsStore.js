import { defineStore } from "pinia";
import { merge } from "lodash";
import { dataPptLastToPrev } from "@/helpers/dataPptLastToPrev.js";
import { dataGetLast } from "@/helpers/dataGetLast.js";
import DatasetsStore_CpiGroups_top from "@/store/DatasetsStore_CpiGroups_top.js";

export const useDatasets = defineStore({
  id: 'datasets',

  state: () => ({
    _ready: false,
    cpi: null,
    cpi_energie: null,
    cpi_voedingsmiddelen: null,
    cpi_groepen: null
  }),

  getters: {
    // CPI
    g_cpi_YoY(state) {
      return dataGetLast(state.cpi);
    },
    g_cpi_PPT(state) {
      return dataPptLastToPrev(state.cpi);
    },

    // CPI energy
    g_cpiEnergy_YoY(state) {
      return dataGetLast(state.cpi_energie);
    },
    g_cpiEnergy_PPT(state) {
      return dataPptLastToPrev(state.cpi_energie);
    },

    // CPI food
    g_cpiFood_YoY(state) {
      return dataGetLast(state.cpi_voedingsmiddelen);
    },
    g_cpiFood_PPT(state) {
      return dataPptLastToPrev(state.cpi_voedingsmiddelen);
    },

    // get highest CPI YoY groups (excl. energy/food)
    g_cpiGroups_top(state) {
      return DatasetsStore_CpiGroups_top(state);
    }
  },

  actions: {
    fetchDatasets() {
      let store = this;
      // fetch dataset from url
      fetch(import.meta.env.VITE_DATASET_URL)
        // parse json response
        .then((response) => response.json())
        .then((data) => {
          if (import.meta.env.VITE_THROTTLE_FETCH) {
            setTimeout(() => {
              merge(store.$state, data);
              store._ready = true;
            }, 1000);
          } else {
            merge(store.$state, data);
            store._ready = true;
          }
        })
        .catch(() => {
          console.error('Could not fetch datasets!');
        });
    },
  },
});