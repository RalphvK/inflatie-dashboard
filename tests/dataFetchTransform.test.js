import { describe, assert, it, expect, toMatchSnapshot } from "vitest";
import { useDatasets } from "../src/store/DatasetsStore";
import { setActivePinia, createPinia } from 'pinia';
import * as datasetsData from './data/dashboard_datasets.json';

import { dataGetLastAsPercentage } from '@/helpers/dataGetLast.js';
import _ from "lodash";
import { getLastPeriodTitle_MY } from "../src/helpers/getLastPeriodTitle";

setActivePinia(createPinia());
const datasetsStore = useDatasets();

describe('datasetsStore', () => {
  
  it('fetchDatasets', () => {
    datasetsStore.mergeDatasets(datasetsData);
    expect(datasetsStore._ready).toEqual(true);
    expect(datasetsStore.cpi).toSatisfy(function(value) {
      if (!Array.isArray(value)) { return false; }
      if (value.length < 12) { return false; }
      return true;
    });
    console.log(datasetsStore.cpi);
  });

  it('getters', () => {
    expect(datasetsStore.g_cpi_YoY).toMatchSnapshot();
    expect(datasetsStore.g_cpi_PPT).toMatchSnapshot();
    expect(datasetsStore.g_cpiEnergy_YoY).toMatchSnapshot();
    expect(datasetsStore.g_cpiEnergy_PPT).toMatchSnapshot();
    expect(datasetsStore.g_cpiFood_YoY).toMatchSnapshot();
    expect(datasetsStore.g_cpiFood_PPT).toMatchSnapshot();
    expect(datasetsStore.g_cpiGroups_top).toMatchSnapshot();
  });

});

describe('getLastPeriodTitle', () => {

  it('getLastPeriodTitle_MY', () => {
    expect(getLastPeriodTitle_MY(datasetsStore.cpi)).toEqual('juli 2022');
  });

});
