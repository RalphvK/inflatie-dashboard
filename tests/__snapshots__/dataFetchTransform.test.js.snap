// Vitest Snapshot v1

exports[`datasetsStore > getters 1`] = `
SentientPercentage {
  "currencySymbol": "€",
  "intValue": 103,
  "percentagePointSymbol": "ppt",
  "percentageSymbol": "%",
  "precision": 1,
  "separator_decimal": ",",
  "separator_thousands": ".",
}
`;

exports[`datasetsStore > getters 2`] = `
SentientPercentagePoint {
  "currencySymbol": "€",
  "intValue": 17,
  "percentagePointSymbol": "ppt",
  "percentageSymbol": "%",
  "precision": 1,
  "separator_decimal": ",",
  "separator_thousands": ".",
}
`;

exports[`datasetsStore > getters 3`] = `
SentientPercentage {
  "currencySymbol": "€",
  "intValue": 1078,
  "percentagePointSymbol": "ppt",
  "percentageSymbol": "%",
  "precision": 1,
  "separator_decimal": ",",
  "separator_thousands": ".",
}
`;

exports[`datasetsStore > getters 4`] = `
SentientPercentagePoint {
  "currencySymbol": "€",
  "intValue": 243,
  "percentagePointSymbol": "ppt",
  "percentageSymbol": "%",
  "precision": 1,
  "separator_decimal": ",",
  "separator_thousands": ".",
}
`;

exports[`datasetsStore > getters 5`] = `
SentientPercentage {
  "currencySymbol": "€",
  "intValue": 123,
  "percentagePointSymbol": "ppt",
  "percentageSymbol": "%",
  "precision": 1,
  "separator_decimal": ",",
  "separator_thousands": ".",
}
`;

exports[`datasetsStore > getters 6`] = `
SentientPercentagePoint {
  "currencySymbol": "€",
  "intValue": 11,
  "percentagePointSymbol": "ppt",
  "percentageSymbol": "%",
  "precision": 1,
  "separator_decimal": ",",
  "separator_thousands": ".",
}
`;
